### ISSUE TITLE
Add Generation of Times

### ISSUE THREAD

**Issue by zinki97** [jlink/jqwik#154]
## Testing Problem

Generating times with certain constraints is a standard problem when building domain-specific
generators. Currently jqwik requires users to generate times from numbers or other primitive data type.

## Suggested Solution

Add arbitraries to the new module/artefact (id "jqwik-time") to generate time values.

Foundation for generation are the types in java.time. Mappers to "old" JDK classes like Date, Timestamp, Calendar etc. should be provided for convenience.

Add a new class to the base package net.jqwik.api.time:

- net.jqwik.api.time.Times

In the long run default generation by type should be possible for all time types, including Duration. There should also be annotations available for the most common time constraints, e.g. before, after, between etc.

**Comment by zinki97** [comment]
### Time Generation Options
* [ ] Min/Max Hours
* [ ] Min/Max Minutes
* [ ] Min/Max Seconds
* [ ] Min/Max Milliseconds
* [ ] Min/Max Nanoseconds
* [ ] Between given times (borders in- or excluded)
* [ ] Constrain precision (nanos, millis, seconds, minutes)
* [ ] Include/exclude lap seconds

### Default Generation of Dates
* [ ] LocalTime
* [ ] Duration

### Annotations for Time
* [ ] `@TimeRange`
* [ ] `@HourRange`
* [ ] `@MinuteRange`
* [ ] `@SecondRange`
* [ ] `@MillisecondRange`
* [ ] `@NanosecondRange`

**Comment by zinki97** [comment]
Design idea:

class `Times`:
- `LocalTimeArbitrary times()`
- `DurationArbitrary durations()`

class `LocalTimeArbitrary`:
TODO

**Comment by jlink** [comment]
I would provide more options in class `Times`:
- `LocalTimeArbitrary localTimes()`
- `OffsetTimeArbitrary offsetTimes()`
- `LocalDateTimeArbitrary localDateTimes()`
- `DateArbitrary localDateTimesAsDate()`
- `InstantArbitrary instants()`
- `OffsetDateTimeArbitrary offsetDateTimes()`
- `ZonedDateTimeArbitrary zonedDateTimes()`

And probably one of:
- `ZoneIdArbitrary zoneIds()`
- `TimeZoneArbitrary timeZones()`
I'm not sure which is more generic and useful.

**Comment by zinki97** [comment]
I think it's better to add DateTimes in a separate step and in this step only Times.

**Comment by jlink** [comment]
> I think it's better to add DateTimes in a separate step and in this step only Times.

Fine with me.

**Comment by jlink** [comment]
OffsetTime would be in-scope then.

**Comment by jlink** [comment]
One aspect that's interesting in times is on the day of DST change from summer to winter time, because a certain hour (usually 2am-3am) can appear twice. I guess OffsetTime can represent that and OffsetTimeArbitrary should be configurable to include/exclude it.

**Comment by zinki97** [comment]
> One aspect that's interesting in times is on the day of DST change from summer to winter time, because a certain hour (usually 2am-3am) can appear twice. I guess OffsetTime can represent that and OffsetTimeArbitrary should be configurable to include/exclude it.

I'll take a closer look at it when I implement OffsetTimeArbitrary.

**Comment by zinki97** [comment]
"Include/exclude leap seconds"

I think that is not possible because the classes does not include `23:59:60`. What do you think?

**Comment by jlink** [comment]
You mean it’s not possible to create a time with 60 seconds? If so forget about leap seconds.

**Comment by zinki97** [comment]
> You mean it’s not possible to create a time with 60 seconds? If so forget about leap seconds.

Yeah, I've testet it with `LocalTime`:

`System.out.println(LocalTime.of(23, 59, 60));`
```
java.time.DateTimeException:
    Invalid value for SecondOfMinute (valid values 0 - 59): 60
```

**Comment by zinki97** [comment]
Maybe we should add a new class `net.jqwik.api.time.DateTimes` for `DateTimes`?

**Comment by jlink** [comment]
Sounds reasonable. But i want to finish and merge all time-related stuff first.

**Comment by zinki97** [comment]
> Sounds reasonable. But i want to finish and merge all time-related stuff first.

I know. It was only consideration

**Comment by jlink** [comment]
This one is done and available in 1.5.1-SNAPSHOT

Follow up issue for DateTimes is https://github.com/jlink/jqwik/issues/175
